lib/newapi1.js

```diff
@@ -0,0 +1,305 @@
+
+    if (options.runScripts === "outside-only") {
+      vm.createContext(this[window]);
+      this[window]._document._defaultView = this[window]._globalProxy = vm.runInContext("this", this[window]);
+    }
+
+    options.beforeParse(this[window]._globalProxy);
+
+    // TODO NEWAPI: this is still pretty hacky. It's also different than jsdom.jsdom. Does it work? Can it be better?
+    documentImpl._htmlToDom.appendHtmlToDocument(html, documentImpl);
+    documentImpl.close();
+  }
+
+  get window() {
+    // It's important to grab the global proxy, instead of just the result of `new Window(...)`, since otherwise things
+    // like `window.eval` don't exist.
+    return this[window]._globalProxy;
+  }
+
+  get virtualConsole() {
+    return this[window]._virtualConsole;
+  }
+
+  get cookieJar() {
+    // TODO NEWAPI move _cookieJar to window probably
+    return idlUtils.implForWrapper(this[window]._document)._cookieJar;
+  }
+
+  serialize() {
+    return domToHtml([this[window]._document]);
+  }
+
+  nodeLocation(node) {
+    if (!idlUtils.implForWrapper(this[window]._document)._parseOptions.locationInfo) {
+      throw new Error("Location information was not saved for this jsdom. Use includeNodeLocations during creation.");
+    }
+
+    return idlUtils.implForWrapper(node)[locationInfo];
+  }
+
+  runVMScript(script) {
+    if (!vm.isContext(this[window])) {
+      throw new TypeError("This jsdom was not configured to allow script running. " +
+        "Use the runScripts option during creation.");
+    }
+
+    return script.runInContext(this[window]);
+  }
+
+  reconfigure(settings) {
+    if ("windowTop" in settings) {
+      this[window]._top = settings.windowTop;
+    }
+
+    if ("url" in settings) {
+      const document = idlUtils.implForWrapper(this[window]._document);
+
+      const url = whatwgURL.parseURL(settings.url);
+      if (url === "failure") {
+        throw new TypeError(`Could not parse "${settings.url}" as a URL`);
+      }
+
+      document._URL = url;
+      document._origin = whatwgURL.serializeURLToUnicodeOrigin(document._URL);
+    }
+  }
+
+  static fragment(string, options) {
+    const dom = new JSDOM(``, options);
+    const template = dom.window.document.createElement("template");
+    template.innerHTML = string;
+    return template.content;
+  }
+
+  static fromURL(url, options) {
+    return Promise.resolve().then(() => {
+      url = (new URL(url)).href;
+      options = normalizeFromURLOptions(options);
+
+      const requestOptions = {
+        resolveWithFullResponse: true,
+        gzip: true,
+        headers: {
+          "User-Agent": options.userAgent,
+          Referer: options.referrer,
+          Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
+          "Accept-Language": "en"
+        },
+        jar: wrapCookieJarForRequest(options.cookieJar)
+      };
+
+      return request(url, requestOptions).then(res => {
```

Comment:
Just return and start a new outer-scope `.then`?

I guess that'd mean pulling out options into the outer scope which makes them throw?